mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -603,8 +603,10 @@ def compute_next_token(current_size):
             parsed_orderby, sorting_joins = _get_orderby_clauses(order_by, session)
 
             query = session.query(SqlRun)
-            for j in _get_sqlalchemy_filter_clauses(parsed_filters, session) + sorting_joins:
+            for j in _get_sqlalchemy_filter_clauses(parsed_filters, session):
                 query = query.join(j)
+            for j in sorting_joins:
```

Comment:
Could you add a comment on why `sorted_joins` for order by columns needs `OUTER JOIN` semantics. Also, will this automatically push the runs that are missing keys in order by columns to the end in sorting order. They should be last regardless of whether order by is `ASC` or `DESC` to match the previous behavior prior to sql filter push downs in #2059 

cc: @dbczumar

Reply:
@rom1504 Can you add a test case confirming that, when ordering by a key, runs without the key are displayed last for both ascending and descending orderings?

@mparkhe I manually confirmed that runs with missing keys in order by columns are returned last in `search_runs()` results. I ran the following logging code:

```
import mlflow
mlflow.set_tracking_uri("sqlite:///dbtest")

for i in range(5):
    with mlflow.start_run():
        mlflow.log_params({
            "param": str(i)
        })
        if i % 2 == 0:
            mlflow.log_metrics({
                "metric": i
            })
```

Then, I performed a search ordered by `metrics.metric asc`:
```
from mlflow.tracking.client import MlflowClient
client = MlflowClient()
asc_results = client.search_runs(experiment_ids=0, order_by=["metrics.metric asc"])
print([result.data for result in asc_results])
```

This produced the following output, confirming that runs for which `metrics.metric` is absent are displayed last:
```
[<RunData: metrics={u'metric': 0.0}, params={u'param': u'0'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={u'metric': 2.0}, params={u'param': u'2'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={u'metric': 4.0}, params={u'param': u'4'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={}, params={u'param': u'3'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={}, params={u'param': u'1'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>]
```

Similarly, I then performed a search ordered by `metrics.metric desc`:
```
from mlflow.tracking.client import MlflowClient
client = MlflowClient()
asc_results = client.search_runs(experiment_ids=0, order_by=["metrics.metric asc"])
print([result.data for result in asc_results])
```

This produced the following result (runs that don't define `metrics.metric` are displayed last):
```
[<RunData: metrics={u'metric': 4.0}, params={u'param': u'4'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={u'metric': 2.0}, params={u'param': u'2'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={u'metric': 0.0}, params={u'param': u'0'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={}, params={u'param': u'3'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>, <RunData: metrics={}, params={u'param': u'1'}, tags={u'mlflow.source.git.commit': u'37f66b28a00e3f75791edb92956d88628f12b5e6',
 u'mlflow.source.name': u'log.py',
 u'mlflow.source.type': u'LOCAL',
 u'mlflow.user': u'czumar'}>]
```

Reply:
That's a good point to add a test case for this !
I added a test case in an existing test for this (with a comment explaining what is the expected order when sorting)

Reply:
Thanks @dbczumar and @rom1504